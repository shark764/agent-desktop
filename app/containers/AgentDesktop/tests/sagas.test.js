import {
  loadHistoricalInteractionBody,
  loadContactInteractions,
  goNotReady,
  goDeleteContacts,
} from 'containers/AgentDesktop/sagas';

describe('loadHistoricalInteractionBody Saga', () => {
  describe('if bodyType is recordings', () => {
    const mockAction = {
      bodyType: 'recordings',
      interactionId: 'mockId',
    };
    const generator = loadHistoricalInteractionBody(mockAction);
    const mockGetRecordings = 'mockGetRecordingsFunction';
    const mockRecordings = [{
      url: 'mockRecordingUrl',
    }];
    beforeEach(() => {
      global.SDK = {
        interactions: {
          voice: {
            getRecordings: mockGetRecordings,
          },
        },
      };
    });
    it('should call the promise util with the SDK getRecordings function with the correct args', () => {
      expect(generator.next()).toMatchSnapshot();
    });
    it('should use the yielded SDK results to dispatch an updateContactHistoryInteractionDetails action with the correct args', () => {
      expect(generator.next(mockRecordings)).toMatchSnapshot();
    });
  });
  describe('if bodyType is transcript', () => {
    const mockAction = {
      bodyType: 'transcript',
      interactionId: 'mockId',
    };
    const generator = loadHistoricalInteractionBody(mockAction);
    const mockGetTranscripts = 'mockGetTranscriptsFunction';
    const mockGetTranscriptsResponse = [{
      url: 'mockTranscriptUrl',
    }];
    const mockTranscriptUrlResponse = {
      data: 'mockTranscript',
    };
    beforeEach(() => {
      global.SDK = {
        interactions: {
          messaging: {
            getTranscripts: mockGetTranscripts,
          },
        },
      };
    });
    it('should call the promise util with the SDK getTranscripts function with the correct args', () => {
      expect(generator.next()).toMatchSnapshot();
    });
    it('should call axios get with the 1st returned transcript url', () => {
      expect(generator.next(mockGetTranscriptsResponse)).toMatchSnapshot();
    });
    it('should use the yielded SDK results to dispatch an updateContactHistoryInteractionDetails action with the correct args', () => {
      expect(generator.next(mockTranscriptUrlResponse)).toMatchSnapshot();
    });
  });
});

describe('loadContactInteractions', () => {
  const mockInteractionHistory = 'mockInteractionHistoryDetails';

  beforeEach(() => {
    global.SDK = {
      reporting: {
        getContactHistory: 'getContactHistory',
      },
    };
  });

  describe('when action.page is undefined', () => {
    const mockAction = {
      contactId: 'mockContactId',
    };
    const generator = loadContactInteractions(mockAction);
    it('should call the promise util with the SDK getContactHistory with the correct arguments', () => {
      expect(generator.next()).toMatchSnapshot();
    });
    it('should use the yielded SDK results to dispatch a setContactInteractionHistory action with the correct args', () => {
      expect(generator.next(mockInteractionHistory));
    });
  });

  describe('when action.page is defined', () => {
    const mockAction = {
      contactId: 'mockContactId',
      page: 1,
    };
    const generator = loadContactInteractions(mockAction);
    it('should call the promise util with the SDK getContactHistory with the correct arguments', () => {
      expect(generator.next()).toMatchSnapshot();
    });
    it('should use the yielded SDK results to dispatch a setContactInteractionHistory action with the correct args', () => {
      expect(generator.next(mockInteractionHistory));
    });
  });
});

describe('goNotReady', () => {
  beforeEach(() => {
    global.SDK = {
      session: {
        goNotReady: 'goNotReady',
      },
    };
  });

  describe('when action.reason is undefined', () => {
    const mockGoNotReadyResponse = 'mockGoNotReadyResponse';
    const mockAction = {};
    const generator = goNotReady(mockAction);
    it('should call the promise util with the SDK goNotReady and the correct arguments', () => {
      expect(generator.next()).toMatchSnapshot();
    });
    it('should use the yielded SDK response to dispatch a setPresence action with the correct args', () => {
      expect(generator.next(mockGoNotReadyResponse));
    });
  });

  describe('when action.reason is defined', () => {
    const mockReasonListId = 'mockReasonListId';
    const mockGoNotReadyResponse = 'mockGoNotReadyResponse';
    const mockAction = {
      listId: 'mockReasonListId',
      reason: {
        name: 'mockReasonName',
        reasonId: 'mockReasonId',
      },
    };
    const generator = goNotReady(mockAction);
    it('should use the yielded id to call the promise util with the SDK goNotReady and the correct arguments', () => {
      expect(generator.next(mockReasonListId)).toMatchSnapshot();
    });
    it('should use the yielded SDK response to dispatch a setPresence action with the correct args', () => {
      expect(generator.next(mockGoNotReadyResponse));
    });
  });
});

describe('goDeleteContacts', () => {
  let generator;
  beforeEach(() => {
    global.SDK = {
      contacts: {
        delete: 'deleteContacts',
      },
    };
  });

  describe('when action.reasonIds is an array of Ids', () => {
    beforeEach(() => {
      const mockAction = {
        contactIds: ['reasonId1', 'reasonId2'],
      };
      generator = goDeleteContacts(mockAction);
    });
    it('should call the promise util with the SDK goDeleteContacts and the correct arguments', () => {
      expect(generator.next()).toMatchSnapshot();
    });
    describe('when API responds with true bools', () => {
      const mockDeleteResponses = [true, true];
      it('should use the yielded SDK responses to dispatch a removeContact action for each contactId', () => {
        generator.next();
        expect(generator.next(mockDeleteResponses)).toMatchSnapshot();
      });
    });
    describe('when API responds with a false bool for contactId1 and a true bool for 2', () => {
      const mockDeleteResponses = [false, true];
      it('should only dispatch a removeContact action contactId2', () => {
        generator.next();
        expect(generator.next(mockDeleteResponses)).toMatchSnapshot();
      });
    });
    it('should dispatch a clearSearchResults action after deletion is complete', () => {
      generator.next();
      generator.next([]);
      expect(generator.next()).toMatchSnapshot();
    });
  });
});
