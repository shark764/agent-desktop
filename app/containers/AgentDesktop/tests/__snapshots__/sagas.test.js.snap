// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`goDeleteContacts when action.reasonIds is an array of Ids should call the promise util with the SDK goDeleteContacts and the correct arguments 1`] = `
Object {
  "done": false,
  "value": Array [
    Object {
      "@@redux-saga/IO": true,
      "CALL": Object {
        "args": Array [
          "deleteContacts",
          Object {
            "contactId": "reasonId1",
          },
          "AgentDesktop",
        ],
        "context": null,
        "fn": [Function],
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "CALL": Object {
        "args": Array [
          "deleteContacts",
          Object {
            "contactId": "reasonId2",
          },
          "AgentDesktop",
        ],
        "context": null,
        "fn": [Function],
      },
    },
  ],
}
`;

exports[`goDeleteContacts when action.reasonIds is an array of Ids should dispatch a clearSearchResults action after deletion is complete 1`] = `
Object {
  "done": false,
  "value": Object {
    "@@redux-saga/IO": true,
    "PUT": Object {
      "action": Object {
        "type": "app/ContactsControl/CLEAR-SEARCH_RESULTS",
      },
      "channel": null,
    },
  },
}
`;

exports[`goDeleteContacts when action.reasonIds is an array of Ids when API responds with a false bool for contactId1 and a true bool for 2 should only dispatch a removeContact action contactId2 1`] = `
Object {
  "done": false,
  "value": Array [
    Object {
      "@@redux-saga/IO": true,
      "PUT": Object {
        "action": Object {
          "contactId": "reasonId2",
          "type": "app/AgentDesktop/REMOVE_CONTACT",
        },
        "channel": null,
      },
    },
  ],
}
`;

exports[`goDeleteContacts when action.reasonIds is an array of Ids when API responds with true bools should use the yielded SDK responses to dispatch a removeContact action for each contactId 1`] = `
Object {
  "done": false,
  "value": Array [
    Object {
      "@@redux-saga/IO": true,
      "PUT": Object {
        "action": Object {
          "contactId": "reasonId1",
          "type": "app/AgentDesktop/REMOVE_CONTACT",
        },
        "channel": null,
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "PUT": Object {
        "action": Object {
          "contactId": "reasonId2",
          "type": "app/AgentDesktop/REMOVE_CONTACT",
        },
        "channel": null,
      },
    },
  ],
}
`;

exports[`goNotReady when action.reason is defined should use the yielded id to call the promise util with the SDK goNotReady and the correct arguments 1`] = `
Object {
  "done": false,
  "value": Object {
    "@@redux-saga/IO": true,
    "CALL": Object {
      "args": Array [
        "goNotReady",
        Object {
          "reasonInfo": Object {
            "reason": "mockReasonName",
            "reasonId": "mockReasonId",
            "reasonListId": "mockReasonListId",
          },
        },
        "AgentDesktop",
      ],
      "context": null,
      "fn": [Function],
    },
  },
}
`;

exports[`goNotReady when action.reason is undefined should call the promise util with the SDK goNotReady and the correct arguments 1`] = `
Object {
  "done": false,
  "value": Object {
    "@@redux-saga/IO": true,
    "CALL": Object {
      "args": Array [
        "goNotReady",
        Object {},
        "AgentDesktop",
      ],
      "context": null,
      "fn": [Function],
    },
  },
}
`;

exports[`loadContactInteractions when action.page is defined should call the promise util with the SDK getContactHistory with the correct arguments 1`] = `
Object {
  "done": false,
  "value": Object {
    "@@redux-saga/IO": true,
    "CALL": Object {
      "args": Array [
        "getContactHistory",
        Object {
          "entityId": "mockContactId",
          "page": 1,
        },
        "AgentDesktop",
      ],
      "context": null,
      "fn": [Function],
    },
  },
}
`;

exports[`loadContactInteractions when action.page is undefined should call the promise util with the SDK getContactHistory with the correct arguments 1`] = `
Object {
  "done": false,
  "value": Object {
    "@@redux-saga/IO": true,
    "CALL": Object {
      "args": Array [
        "getContactHistory",
        Object {
          "entityId": "mockContactId",
        },
        "AgentDesktop",
      ],
      "context": null,
      "fn": [Function],
    },
  },
}
`;

exports[`loadHistoricalInteractionBody Saga if bodyType is recordings should call the promise util with the SDK getRecordings function with the correct args 1`] = `
Object {
  "done": false,
  "value": Object {
    "@@redux-saga/IO": true,
    "CALL": Object {
      "args": Array [
        "mockGetRecordingsFunction",
        Object {
          "interactionId": "mockId",
        },
        "AgentDesktop",
      ],
      "context": null,
      "fn": [Function],
    },
  },
}
`;

exports[`loadHistoricalInteractionBody Saga if bodyType is recordings should use the yielded SDK results to dispatch an updateContactHistoryInteractionDetails action with the correct args 1`] = `
Object {
  "done": false,
  "value": Object {
    "@@redux-saga/IO": true,
    "PUT": Object {
      "action": Object {
        "interactionDetails": Object {
          "audioRecordings": Array [
            "mockRecordingUrl",
          ],
        },
        "interactionId": "mockId",
        "type": "app/AgentDesktop/UPDATE_CONTACT_HISTORY_INTERACTION_DETAILS",
      },
      "channel": null,
    },
  },
}
`;

exports[`loadHistoricalInteractionBody Saga if bodyType is transcript should call axios get with the 1st returned transcript url 1`] = `
Object {
  "done": false,
  "value": Object {
    "@@redux-saga/IO": true,
    "CALL": Object {
      "args": Array [
        "mockTranscriptUrl",
      ],
      "context": null,
      "fn": [Function],
    },
  },
}
`;

exports[`loadHistoricalInteractionBody Saga if bodyType is transcript should call the promise util with the SDK getTranscripts function with the correct args 1`] = `
Object {
  "done": false,
  "value": Object {
    "@@redux-saga/IO": true,
    "CALL": Object {
      "args": Array [
        "mockGetTranscriptsFunction",
        Object {
          "interactionId": "mockId",
        },
        "AgentDesktop",
      ],
      "context": null,
      "fn": [Function],
    },
  },
}
`;

exports[`loadHistoricalInteractionBody Saga if bodyType is transcript should use the yielded SDK results to dispatch an updateContactHistoryInteractionDetails action with the correct args 1`] = `
Object {
  "done": false,
  "value": Object {
    "@@redux-saga/IO": true,
    "PUT": Object {
      "action": Object {
        "interactionDetails": Object {
          "transcript": "mockTranscript",
        },
        "interactionId": "mockId",
        "type": "app/AgentDesktop/UPDATE_CONTACT_HISTORY_INTERACTION_DETAILS",
      },
      "channel": null,
    },
  },
}
`;
